src/t_list.c

```diff
@@ -125,17 +112,31 @@ void listTypeTryConvertQuicklist(robj *o, int shrinking, beforeConvertCB fn, voi
     /* Invoke callback before conversion. */
     if (fn) fn(data);
 
-    listTypeConvertToListpack(o);
+    /* Extract the listpack from the unique quicklist node,
+     * then reset it and release the quicklist. */
+    o->ptr = ql->head->entry;
+    ql->head->entry = NULL;
+    quicklistRelease(ql);
+    o->encoding = OBJ_ENCODING_LISTPACK;
 }
 
-void listTypeTryConversionForShrinking(robj *o, beforeConvertCB fn, void *data) {
-    if (o->encoding == OBJ_ENCODING_QUICKLIST) {
-        listTypeTryConvertQuicklist(o, 1, fn, data);
-    } else if (o->encoding == OBJ_ENCODING_LISTPACK) {
-        /* Nothing to do */
-    } else {
+/* Check if the list needs to be converted to appropriate encoding due to
+ * growing, shrinking or other cases.
+ *
+ * 'lct' can be one of the following values:
+ * LIST_CONV_GROWING   - Only consider converting from listpack to quicklist
+ * LIST_CONV_SHRINKING - Only consider converting from quicklist to listpack
+ * LIST_CONV_UNKNOWN   - Let system decide whether to convert or not */
+void listTypeTryConversion(robj *o, list_conv_type lct,
```

Comment:
let's also document the argv thing.
btw, i don't mind adding an alias for this function without these argument, it'll still be clear.
i.e. calling it with data that's going to be appended.

Reply:
Done, I add a new interface `listTypeTryConversionWithArgv`, not sure if it's good enought.